sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/Misc/MiscellaneousTests.cs

```diff
@@ -0,0 +1,245 @@
+        [TestCategory("Slow")]
+        public void TestForSlowAnalyzers_EFMigrationSource_2474()
+        {
+            // The test file comes from a repro project for https://github.com/SonarSource/sonar-dotnet/issues/2474
+            // The single file is enough to cause the issue described in the bug, although there
+            // are other large files in the project that might also be problematic.
+
+            // Analyzers known to timeout against this test code
+            // (note: this test doesn't cover the utility analyzers):
+            var knownSlowAnalyzers = new[] { "SonarAnalyzer.Rules.CSharp.UnusedPrivateMember" };
+            
+            int analysisTimeoutInMs = 700;
+            var allAnalyzerTypes = GetSonarAnalyzerTypes(typeof(MetricsAnalyzer).Assembly);
+            var compilation = GetEFCompilation();
+
+            var executionResults = ExecuteAllAnalyzers(compilation, allAnalyzerTypes, analysisTimeoutInMs);
+
+            AssertNoUnexpectedSlowAnalyzers(executionResults, knownSlowAnalyzers);
+        }
+
+        /// <summary>
+        /// This method runs each of the supplied C# analyzers in series against the compilation.
+        /// Any analyzer that takes longer than a specifed timeout is reported and that analysis is stopped.
+        /// NOTE: the utility analyzers won't run because the required analysis settings are not set
+        /// by the test.
+        /// </summary>
+        private IEnumerable<ExecutionResult> ExecuteAllAnalyzers(Compilation compilation, IEnumerable<Type> analyzerTypes, int analysisTimeoutInMs)
+        {
+            // Run an initial analysis so any startup costs aren't attributed to the first analyzer to be tested
+            ExecuteAnalyzerWithTimeout(compilation, CreateAnalyzer(analyzerTypes.First()), 5000);
+
+            var executionResults = new List<ExecutionResult>();
+            var counter = 0;
+            var timer = new Stopwatch();
+
+            // Executing one analyzer at a time (running them in parallel can skew the results)
+            foreach (var analyzerType in analyzerTypes)
+            {
+                var analyzer = CreateAnalyzer(analyzerType);
+
+                // Write to the DEBUG window so we can see what's happening in the output window in the IDE
+                counter++;
+                Debug.WriteLine($"{counter}: {analyzerType.FullName}");
+
+                timer.Restart();
+                var succeeded = ExecuteAnalyzerWithTimeout(compilation, analyzer, analysisTimeoutInMs);
+                timer.Stop();
+
+                var newExecutionResult = new ExecutionResult(analyzerType, timer.ElapsedMilliseconds, !succeeded);
+                executionResults.Add(newExecutionResult);
+
+                if (!succeeded)
+                {
+                    Debug.WriteLine($"   -> TIMED OUT");
+                }
+            }
+
+            DumpExecutionResults(executionResults);
```

Comment:
Writes the execution results to the test output, sorted by execution time (descending).